mitmproxy/events.py

```diff
@@ -0,0 +1,85 @@
+import re
+from dataclasses import dataclass, is_dataclass, fields
+from typing import ClassVar, Any, Dict, Type, Set, List, TYPE_CHECKING, Sequence
+
+import mitmproxy.flow
+
+if TYPE_CHECKING:
+    import mitmproxy.addonmanager
+    import mitmproxy.log
+
+
+class MitmproxyEvent:
+    name: ClassVar[str]
+
+    def args(self) -> List[Any]:
+        args = []
+        for field in fields(self):
+            args.append(getattr(self, field.name))
+        return args
+
+    def __new__(cls, *args, **kwargs):
+        if cls is MitmproxyEvent:
+            raise TypeError("MitmproxyEvent may not be instantiated directly.")
+        if not is_dataclass(cls):
+            raise TypeError("Subclass is not a dataclass.")
+        return super().__new__(cls)
+
+    def __init_subclass__(cls, **kwargs):
+        # initialize .name attribute. HttpRequestHook -> http_request
+        if cls.__dict__.get("name", None) is None:
+            name = cls.__name__.replace("Hook", "").replace("Event", "")
+            cls.name = re.sub('(?!^)([A-Z]+)', r'_\1', name).lower()
+        if cls.name in all_events:
+            other = all_events[cls.name]
+            raise RuntimeError(f"Two conflicting event classes for {cls.name}: {cls} and {other}")
+        if cls.name == "":
+            return  # don't register Hook class.
+        all_events[cls.name] = cls
+
+        # define a custom hash and __eq__ function so that events are hashable and not comparable.
+        cls.__hash__ = object.__hash__
+        cls.__eq__ = object.__eq__
+
+
+all_events: Dict[str, Type[MitmproxyEvent]] = {}
+
+
+@dataclass
+class ConfigureEvent(MitmproxyEvent):
```

Comment:
I not sure I particularly like the choice of naming strategy:
* `...Event` suffix is redundant - we already know these are events from the import path
* events are something that have happened - meaning past tense
* unless we have a hierarchy or other grouping, they should be descriptive

IMO this matches what h2 is doing: https://github.com/python-hyper/hyper-h2/blob/master/src/h2/events.py
and to some degree also wsproto: https://github.com/python-hyper/wsproto/blob/master/src/wsproto/events.py (although wsproto is considering this "active tense", because events in wsproto are also used to produce output)

For example, I would propose:
* `ConfigureEvent` -> `ConfigurationChanged`
* `DoneEvent` -> `EventLoopTerminated`
* `RunningEvent` -> `StartupCompleted`
* `UpdateEvent` -> `FlowChanged`

(just proposals to demonstrate my idea - feel free to suggest different ones)

Reply:
Right, I'm not super happy with the naming either so far. Some more notes:

1. The event name/name of the addon function is currently automatically derived from the class name (FooBarEvent -> foo_bar method in the addon), which is why they are named the way they are. I'd suggest we keep this general approach, it makes things rather obvious. Of course we could drop the "Event" or replace it with something else.
2. `...Event` is not always redundant, for example `log.LogEvent`. FWIW I initially had them all in mitmproxy/events.py, but that led to really ugly circular imports. All proxy core hooks (which subclass from Hook, which subclasses from MitmproxyEvent) are spread in the individual protocol layers already anyways.
3. I went with event here because that's what we've called them in the docs so far. Do we have a better idea? Maybe calling them `Hook` instead? 

**Pro:** 
  - Consistency with the proxy core
  - maybe conveys that you can change stuff here (compared to reacting to an event)
  
**Con:** 
  - Change an existing established name
  - Hook is a bit less mainstream/obvious/easy than "event"

If you have other ideas, please shoot!

Reply:
Seems like a tricky one indeed.

About Hooks and Events: whats you current definition? Are they the same? Is a subset "user facing" and the others are more "proxy core"? One "bubble" up, the others "down"? There were also Commands...? I think we might need a syllabus soon :-P

About the event name + function name: that is indeed a nice pattern. I still don't like the class names though. Just throwing a bigger thought at it: did you consider ditching the "addons with functions" pattern and adopt a "receiving an object" style that we see with wsproto/h2/h11? We could provide a default dispatch table that would call the "legacy functions", or our users can use the "new addon API" by working with the new objects directly (and overriding the default dispatch table or function with their own)...
Another option, but less preferred even by me, would be to split these two things: Class name (with the style I proposed above), and "Addon API function name". More complexity and potential for confusion though... but a clean separation.

About the LogEvent: nitpick or technicality is that we have "log" as verb and as a noun: `LogLineAdded`, or `AddLogEvent` or `AddLogEntry`?

Reply:
> About Hooks and Events: whats you current definition? Are they the same? Is a subset "user facing" and the others are more "proxy core"? One "bubble" up, the others "down"? There were also Commands...? I think we might need a syllabus soon :-P

The proxy core uses a fairly consistent model:

 - Events come "from the outside" and are generally unpredictable. For example, we have the "DataReceived" event, which signals the arrival of bytes on the network.
 - Commands are (immediate) deterministic reactions to events. The are emitted from the proxy core. For example, the "SendData" command instructs the I/O handler to send out some bytes.
 - A Hook is a Command. For example, we may receive some bytes for an HTTP request (a DataReceived event). The proxy core then decides that `request()` needs to be called, so it emits a <s>`Hook`</s>`HookStart` command. When the command has been finished (this is again happening "outside" and is unpredictable), the I/O handler will feed a <s>`HookReply`</s>`HookCompleted` _event_  into the proxy core.

TL;DR: Events are passed into the proxy core, commands come out of the proxy core. "Go call the request method on all addons" is a command (`HookStart`); "we have finished doing that" is an event (`HookCompleted`).

Events "bubble into the core", commands "bubble out". 😛 

> Just throwing a bigger thought at it: did you consider ditching the "addons with functions" pattern and adopt a "receiving an object" style that we see with wsproto/h2/h11?

That would mean that most addons are essential one function with a giant if-else statement. I don't think that is really what we want, no?

> I still don't like the class names though.

Right. I have just pushed a revised version to unify everything a bit. In a nutshell:

- `MitmproxyEvent` -> `EventHook`
- All new subclasses accordingly. For example, the `configure()` method is now represented as `ConfigureEventHook`. We can also take out "Event" entirely if you prefer.
- In the proxy core: 
  - `Hook` -> `StartHook(Command, EventHook)`
  - `HookReply` -> `HookCompleted(Event)`
- In the (user-facing) docs, we call them "Event Hooks" now. I feel that's a fairly understandable term, what do you think?